File: extension/src/main/java/com/thornbirds/component/ComponentController.java
package com.thornbirds.component;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Basic definition of Component Controller for Android platform.
 *
 * @author YangLi yanglijd@gmail.com
 */
public abstract class ComponentController implements IEventController {
    protected static final boolean DEBUG = true;

    protected final String TAG = getTAG();

    protected abstract String getTAG();

    @NonNull
    private final IEventController mEventController;

    protected final void LogE(String msg) {
        Log.e(TAG, msg);
    }

    public ComponentController(@Nullable IEventController eventController) {
        if (eventController != null) {
            this.mEventController = eventController;
        } else {
            this.mEventController = new EventController() {
                @Override
                protected void LogE(String msg) {
                    throw new RuntimeException(msg);
                }
            };
        }
    }

    @Override
    public final void registerObserverForEvent(int event, IEventObserver observer) {
        mEventController.registerObserverForEvent(event, observer);
    }

    @Override
    public final void unregisterObserverForEvent(int event, IEventObserver observer) {
        mEventController.unregisterObserverForEvent(event, observer);
    }

    @Override
    public final void unregisterObserver(IEventObserver observer) {
        mEventController.unregisterObserver(observer);
    }

    @Override
    public final boolean postEvent(int event) {
        return mEventController.postEvent(event);
    }

    @Override
    public final boolean postEvent(int event, IParams params) {
        return mEventController.postEvent(event, params);
    }

    @Override
    public final void release() {
        mEventController.release();
    }

    protected abstract class Action implements IEventObserver {

        protected Action() {
        }

        protected final void registerAction(int event) {
            registerObserverForEvent(event, this);
        }

        protected final void unregisterAction(int event) {
            unregisterObserverForEvent(event, this);
        }

        protected final void register() {
            onRegister();
        }

        protected final void unregister() {
            unregisterObserver(this);
        }

        protected abstract void onRegister();

        @Override
        public abstract boolean onEvent(int event, IParams params);
    }
}

File: extension/src/main/java/com/thornbirds/component/ComponentController.kt
package com.thornbirds.component

import android.util.Log

/**
 * Basic definition of Component Controller for Android platform.
 *
 * @author YangLi yanglijd@gmail.com
 */
abstract class ComponentController(
    eventController: IEventController?
) : IEventController {

    protected abstract val TAG: String

    private val mEventController: IEventController = eventController ?: object : EventController() {
        override fun logE(msg: String) {
            throw RuntimeException(msg)
        }
    }

    protected fun logE(msg: String) {
        Log.e(TAG, msg)
    }

    override fun registerObserverForEvent(event: Int, observer: IEventObserver) {
        mEventController.registerObserverForEvent(event, observer)
    }

    override fun unregisterObserverForEvent(event: Int, observer: IEventObserver) {
        mEventController.unregisterObserverForEvent(event, observer)
    }

    override fun unregisterObserver(observer: IEventObserver) {
        mEventController.unregisterObserver(observer)
    }

    override fun postEvent(event: Int, params: IParams?): Boolean {
        return mEventController.postEvent(event, params)
    }

    override fun release() {
        mEventController.release()
    }

    protected abstract inner class Action : IEventObserver {
        protected fun registerAction(event: Int) {
            registerObserverForEvent(event, this)
        }

        protected fun unregisterAction(event: Int) {
            unregisterObserverForEvent(event, this)
        }

        protected fun register() {
            onRegister()
        }

        protected fun unregister() {
            unregisterObserver(this)
        }

        protected abstract fun onRegister()

        abstract override fun onEvent(event: Int, params: IParams?): Boolean
    }
}